src/analysis/processing/qgsalgorithmvoronoipolygons.cpp

```diff
@@ -0,0 +1,194 @@
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QgsFeatureRequest request;
+  if ( !copyAttributes )
+  {
+    request.setSubsetOfAttributes( QList< int >() );
+  }
+  QgsFeatureIterator it = source->getFeatures( request, QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+
+  QgsGeometry allPoints;
+  QHash< QgsFeatureId, QgsAttributes > attributeCache;
+
+  long i = 0;
+  double step = source->featureCount() > 0 ? 50.0 / source->featureCount() : 1;
+
+  const QgsSpatialIndex index( it, [&]( const QgsFeature & f )->bool
+  {
+    i++;
+    if ( feedback->isCanceled() )
+      return false;
+
+    feedback->setProgress( i * step );
+
+    if ( !f.hasGeometry() )
+      return true;
+
+    const QgsAbstractGeometry *geom = f.geometry().constGet();
+    if ( QgsWkbTypes::isMultiType( geom->wkbType() ) )
+    {
+      const QgsMultiPoint mp( *qgsgeometry_cast< const QgsMultiPoint * >( geom ) );
+      for ( auto pit = mp.const_parts_begin(); pit != mp.const_parts_end(); ++pit )
+      {
+        allPoints.addPart( qgsgeometry_cast< QgsPoint * >( *pit )->clone(), Qgis::GeometryType::Point );
+      }
+    }
+    else
+    {
+      allPoints.addPart( qgsgeometry_cast< QgsPoint * >( geom )->clone(), Qgis::GeometryType::Point );
+    }
+
+    attributeCache.insert( f.id(), f.attributes() );
+
+    return true;
+  }, QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+  QgsRectangle extent = source->sourceExtent();
+  double delta = extent.width() * buffer / 100.0;
+  extent.setXMinimum( extent.xMinimum() - delta );
+  extent.setXMaximum( extent.xMaximum() + delta );
+  delta = extent.height() * buffer / 100.0;
+  extent.setYMinimum( extent.yMinimum() - delta );
+  extent.setYMaximum( extent.yMaximum() + delta );
+  QgsGeometry clippingGeom = QgsGeometry::fromRect( extent );
+
+  QgsGeometry voronoiDiagram = allPoints.voronoiDiagram( clippingGeom, tolerance );
+
+  if ( !voronoiDiagram.isEmpty() )
+  {
+    std::unique_ptr< QgsGeometryEngine > engine;
+    std::unique_ptr< QgsGeometryEngine > extentEngine( QgsGeometry::createGeometryEngine( clippingGeom.constGet() ) );
+    QVector< QgsGeometry > collection = voronoiDiagram.asGeometryCollection();
+    for ( int i = 0; i < collection.length(); i++ )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+      QgsFeature f;
+      f.setFields( fields );
+      f.setGeometry( QgsGeometry( extentEngine->intersection( collection[i].constGet() ) ) );
+      if ( copyAttributes )
+      {
+        const QList< QgsFeatureId > intersected = index.intersects( collection[i].boundingBox() );
+        engine.reset( QgsGeometry::createGeometryEngine( collection[i].constGet() ) );
+        engine->prepareGeometry();
+        for ( const QgsFeatureId id : intersected )
+        {
+          if ( engine->intersects( index.geometry( id ).constGet() ) )
```

Comment:
This should be implied when querying the index built using single points -- we could save the cost here and also avoid storing the geometries in the index.

Potentially a big speed boost by using QgsSpatialIndexKDBush here too.

Reply:
Not sure I understand. We can't check for exact intersection with the geometry, index supports only intersection with the rectangle, so we need test whether point actually intersets geometry.